### PR TITLE
Flask set-cookie codemod

### PR DESCRIPTION
**Pull request by clavedeluna**
## Overview
*New codemod to check if flask's `set_cookie` is called with the most secure params*

## Description

* This codemod is merge after cursory review because it is possible to call `set_cookie` with default (seemingly unsafe) parameters but the safety is configured at the Flask app level configurations. However, whatever is passed to `set_cookie` takes precedence so calling it with unsafe params should be changed.
* Checking flask cookie config settings will be a separate codemod.

### CODE REVIEW

File: tests/codemods/test_secure_flask_cookie.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                "secure=True",
+                "secure=True, httponly=True, samesite='Lax'",
+            ),
+            (
+                "secure=True, httponly=True, samesite='Lax'",
+                "secure=True, httponly=True, samesite='Lax'",
+            ),
+            (
+                "secure=True, httponly=True, samesite='Strict'",
+                "secure=True, httponly=True, samesite='Lax'",
```
Comment (clavedeluna) [inline tests/codemods/test_secure_flask_cookie.py]: We could consider allowing users to pass samesite=Strict and not changing it to Lax

**Review by andrecsilva — APPROVED** [review]
Lacking tests with `open(...)` as a source, otherwise ok.

**Comment by clavedeluna** [comment]
> Lacking tests with `open(...)` as a source, otherwise ok.

Can you clarify with a code example?

**Comment by andrecsilva** [comment]
> > Lacking tests with `open(...)` as a source, otherwise ok.
> 
> Can you clarify with a code example?

Your semgrep rule has 3 possibilities for source (`flask.make_response(...)`, `flask.Response(...)` and `open(...)`). See:
```
            pattern-sources:
              - pattern-either:
                  - patterns:
                    - pattern: flask.make_response(...)
                    - pattern-inside: |
                        import flask
                        ...
                  - patterns:
                    - pattern: flask.Response(...)
                    - pattern-inside: |
                        import flask
                        ...
                  - pattern: open(...)
```

All the unit tests are parameterized with this:
```python
each_func = pytest.mark.parametrize("func", ["make_response", "Response"])
```

Thus, no test covers an example like:
```python
import flask
response = open(something)
var = "hello"
response.set_cookie("name", "value")
```

I'm not sure if this is actually valid code. Just following the semgrep rule.

**Comment by clavedeluna** [comment]
> > > Lacking tests with `open(...)` as a source, otherwise ok.
> > 
> > 
> > Can you clarify with a code example?
> 
> Your semgrep rule has 3 possibilities for source (`flask.make_response(...)`, `flask.Response(...)` and `open(...)`). See:
> 
> ```
>             pattern-sources:
>               - pattern-either:
>                   - patterns:
>                     - pattern: flask.make_response(...)
>                     - pattern-inside: |
>                         import flask
>                         ...
>                   - patterns:
>                     - pattern: flask.Response(...)
>                     - pattern-inside: |
>                         import flask
>                         ...
>                   - pattern: open(...)
> ```
> 
> All the unit tests are parameterized with this:
> 
> ```python
> each_func = pytest.mark.parametrize("func", ["make_response", "Response"])
> ```
> 
> Thus, no test covers an example like:
> 
> ```python
> import flask
> response = open(something)
> var = "hello"
> response.set_cookie("name", "value")
> ```
> 
> I'm not sure if this is actually valid code. Just following the semgrep rule.

OHHHH this is a total typo, from copy paste. open is not a valid source. I will remove it. Good catch!

File: tests/codemods/test_secure_flask_cookie.py (hunk 1 above)
Comment (drdavella) [reply tests/codemods/test_secure_flask_cookie.py]: Yes I think we should not modify `samesite` if it is already less permissive than our codemod.

**Review by drdavella — CHANGES_REQUESTED** [review]
This looks really nice to me overall, but I do think we should preserve `samesite="Strict"` if it's already there. We don't want to make the app less permissive, only moreso.

File: src/core_codemods/docs/pixee_python_secure-flask-cookie.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+This codemod sets the most secure parameters when Flask applications call `set_cookie` on a response object. Without these parameters, your Flask
+application cookies may be vulnerable to being intercepted and used against your system.
```
Comment (drdavella) [inline src/core_codemods/docs/pixee_python_secure-flask-cookie.md]: Just a minor suggestion.
```suggestion
application cookies may be vulnerable to being intercepted and used to gain access to sensitive data.
```

**Review by drdavella — APPROVED** [review]

**Pull request merged**